pkg/controller/operators/operatorconditiongenerator_controller.go

```diff
@@ -170,3 +179,100 @@ func (r *OperatorConditionGeneratorReconciler) ensureOperatorCondition(operatorC
 	existingOperatorCondition.Spec.ServiceAccounts = operatorCondition.Spec.ServiceAccounts
 	return r.Client.Update(context.TODO(), existingOperatorCondition)
 }
+
+// Return values, err, ok; ok == true: continue Reconcile, ok == false: exit Reconcile
+func (r *OperatorConditionGeneratorReconciler) processFinalizer(ctx context.Context, csv *operatorsv1alpha1.ClusterServiceVersion) (error, bool) {
+	myFinalizerName := "operators.coreos.com/csv-cleanup"
+	log := r.log.WithValues("name", csv.GetName()).WithValues("namespace", csv.GetNamespace())
+
+	if csv.ObjectMeta.DeletionTimestamp.IsZero() {
+		// CSV is not being deleted, add finalizer if not present
+		if !controllerutil.ContainsFinalizer(csv, myFinalizerName) {
+			patch := client.MergeFrom(csv.DeepCopy())
+			controllerutil.AddFinalizer(csv, myFinalizerName)
+			if err := r.Client.Patch(ctx, csv, patch); err != nil {
+				log.Error(err, "Adding finalizer")
+				return err, false
+			}
+		}
+		return nil, true
+	}
+
+	if !controllerutil.ContainsFinalizer(csv, myFinalizerName) {
+		// Finalizer has been removed; stop reconciliation as the CSV is being deleted
+		return nil, false
+	}
+
+	// CSV is being deleted and the finalizer still present; do any clean up
+	ownerSelector := ownerutil.CSVOwnerSelector(csv)
+	listOptions := client.ListOptions{
+		LabelSelector: ownerSelector,
+	}
+	deleteOptions := client.DeleteAllOfOptions{
+		ListOptions: listOptions,
+	}
+	// Look for resources owned by this CSV, and delete them.
+	log.WithValues("selector", ownerSelector).Info("Cleaning up resources after CSV deletion")
+	var errs []error
+
+	err := r.Client.DeleteAllOf(ctx, &rbacv1.ClusterRoleBinding{}, &deleteOptions)
+	if client.IgnoreNotFound(err) != nil {
+		log.Error(err, "Deleting ClusterRoleBindings on CSV delete")
+		errs = append(errs, err)
+	}
+
+	err = r.Client.DeleteAllOf(ctx, &rbacv1.ClusterRole{}, &deleteOptions)
+	if client.IgnoreNotFound(err) != nil {
+		log.Error(err, "Deleting ClusterRoles on CSV delete")
+		errs = append(errs, err)
+	}
+
+	err = r.Client.DeleteAllOf(ctx, &admissionregistrationv1.MutatingWebhookConfiguration{}, &deleteOptions)
+	if client.IgnoreNotFound(err) != nil {
+		log.Error(err, "Deleting MutatingWebhookConfigurations on CSV delete")
+		errs = append(errs, err)
+	}
+
+	err = r.Client.DeleteAllOf(ctx, &admissionregistrationv1.ValidatingWebhookConfiguration{}, &deleteOptions)
+	if client.IgnoreNotFound(err) != nil {
+		log.Error(err, "Deleting ValidatingWebhookConfigurations on CSV delete")
+		errs = append(errs, err)
+	}
+
+	// Make sure things are deleted
+	crbList := &rbacv1.ClusterRoleBindingList{}
+	if _ = r.Client.List(ctx, crbList, &listOptions); len(crbList.Items) != 0 {
```

Comment:
We need to use the lister for this, not a live client.

Reply:
"need" or "highly desirable"? There's no code in place for informers/indexers/listers in this part of the code. I'll have to pull in a lot more infrastructure.

Reply:
Need, with a capital N. The server load doing live lists is extremely hard. This is why I made this comment d458b66414

Reply:
OK, to be clear, it's not needed for _functionality_ (as it works already), but needed for _performance_. CSV deletion doesn't occur very often, but it could hit the API server fairly hard if deleting these resources takes longer than expected.

Reply:
Not sure the distinction means much, to be honest. It's needed to meet the quality bar we have for controllers in OCP. Full-cluster resource lists are a good way to DoS the cluster.

Reply:
OK, before I dive too deep into this, let's make sure it's not already using a cache.
The client being used comes from https://github.com/operator-framework/operator-lifecycle-manager/blob/920a92140b3aeb581c48b03635839406ec9624cd/cmd/olm/manager.go#L114

Which  comes from a Controller-Runtime Manager, which is created with a Cache:
https://github.com/operator-framework/operator-lifecycle-manager/blob/920a92140b3aeb581c48b03635839406ec9624cd/cmd/olm/manager.go#L55

AFAICT, the Client from the Manager uses the provided cache. In such a case, I shouldn't have to add separate informers, and should just have to add the appropriate resources to the cache.

https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#Options

Reply:
That's what I thought. Thanks!